### PR TITLE
Keyboard dialog: add password reveal button

### PR DESCRIPTION
**Pull request by ronie**
## Description

i came across a feature request from @scott967 on the forum and thought it made sense.

this PR will add a 'show password' button to the keyboard dialog when the user is asked for a password.

## Motivation and context

it's not particularly easy to enter a long/complex password using a remote control.

## How has this been tested?

tested using the estuary skin and double-checked to new button is hidden in all cases except when a password needs to be entered.

## What is the effect on users?

users can now verify they entered the correct password before clicking the OK button.

## Screenshots (if appropriate):
show

hide

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [X] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by enen92 — APPROVED** [review]
Nice new feature, I've also found this missing sometimes. My only comment is the order of the button on the list. I'd personally move it to the third place on the list (definitely with higher priority w.r.t. the IP address button)

File: addons/resource.language.en_gb/resources/strings.po [hunk 1]
```diff
@@ -5618,7 +5618,17 @@ msgctxt "#12022"
 msgid "Resume from {0:s}"
 msgstr ""
 
-#empty strings from id 12023 to 12309
+#empty strings from id 12023 to 12307
+
+#: xbmc/dialogs/GUIDialogKeyboardGeneric.cpp
```
Comment (ksooo) [inline addons/resource.language.en_gb/resources/strings.po:5623]: could you please also add a verbal description explaining purpose and context of the new string to help the translators?

File: xbmc/dialogs/GUIDialogKeyboardGeneric.cpp [hunk 2]
```diff
@@ -546,6 +557,23 @@ void CGUIDialogKeyboardGeneric::OnSymbols()
...
+    CGUIMessage msg(GUI_MSG_SET_TYPE, GetID(), CTL_EDIT, CGUIEditControl::INPUT_TYPE_PASSWORD);
+    OnMessage(msg);
+  }
+  else
+  {
+    SET_CONTROL_LABEL(CTL_BUTTON_REVEAL, g_localizeStrings.Get(12309));
+    CGUIMessage msg(GUI_MSG_SET_TYPE, GetID(), CTL_EDIT, CGUIEditControl::INPUT_TYPE_TEXT);
+    OnMessage(msg);
+  }
+}
```
Comment (ksooo) [inline xbmc/dialogs/GUIDialogKeyboardGeneric.cpp:575]: May I suggest to optimize this a bit:

```c++
void CGUIDialogKeyboardGeneric::OnReveal()
{
  m_hiddenInput = !m_hiddenInput;
  SET_CONTROL_LABEL(CTL_BUTTON_REVEAL, g_localizeStrings.Get(m_hiddenInput ? 12308 : 12309));
  CGUIMessage msg(GUI_MSG_SET_TYPE, GetID(), CTL_EDIT, m_hiddenInput ? CGUIEditControl::INPUT_TYPE_PASSWORD : CGUIEditControl::INPUT_TYPE_TEXT);
  OnMessage(msg);
}
```

**Review by ksooo — COMMENTED** [review]
Nice.

**Comment by matthuisman** [comment]
So good! 

By the way -
Has anyone thought of adding the @ symbol back to the initial keyboard? (Not via uppercase).
I've been asked a few times where is it. Maybe it could be with the = sign. Quite often used when logging in to addons

**Comment by ronie** [comment]
@enen92 sure, no problem. it's now located below the Cancel button.

@ksooo thanx for the hint. adjusted the code according to your suggestions.

**Review by garbear — APPROVED** [review]
Nice, simple feature. Don't see any problems. Something I've wanted too. Thanks

**Pull request merged**